lib/Sema/NameBinding.cpp

```diff
@@ -36,60 +37,539 @@
+  if (!I.checkModuleLoaded(M, SF)) {
+    // Can't process further. checkModuleLoaded() will have diagnosed this.
+    if (I.ID)
+      I.ID.get()->setModule(nullptr);
+    return;
+  }
+
+  auto topLevelModule = I.getTopLevelModule(M, SF);
+
+  I.validateOptions(topLevelModule, SF);
+
+  if (topLevelModule && topLevelModule != M) {
+    // If we have distinct submodule and top-level module, add both, disabling
+    // the top-level module's scoped import validation.
+    addImport(I, M, false);
+    addImport(I, topLevelModule.get(), true);
+  }
+  else {
+    // Add only the import itself.
+    addImport(I, M, true);
+  }
+
+  if (I.ID)
+    I.ID.get()->setModule(M);
+}
+
+void NameBinder::addImport(const UnboundImport &I, ModuleDecl *M,
+                           bool needsScopeValidation) {
+  auto importDesc = I.makeDesc(M);
+
+  unvalidatedImports.emplace_back(I, importDesc, M, needsScopeValidation);
+}
+
+//===----------------------------------------------------------------------===//
+// MARK: Import module loading
+//===----------------------------------------------------------------------===//
+
+ModuleDecl *NameBinder::getModule(ArrayRef<Located<Identifier>> modulePath) {
   assert(!modulePath.empty());
   auto moduleID = modulePath[0];
-  
+
   // The Builtin module cannot be explicitly imported unless we're a .sil file
   // or in the REPL.
   if ((SF.Kind == SourceFileKind::SIL || SF.Kind == SourceFileKind::REPL) &&
-      moduleID.Item == Context.TheBuiltinModule->getName())
-    return Context.TheBuiltinModule;
+      moduleID.Item == ctx.TheBuiltinModule->getName())
+    return ctx.TheBuiltinModule;
 
   // If the imported module name is the same as the current module,
   // skip the Swift module loader and use the Clang module loader instead.
   // This allows a Swift module to extend a Clang module of the same name.
   //
   // FIXME: We'd like to only use this in SIL mode, but unfortunately we use it
-  // for our fake overlays as well.
+  // for clang overlays as well.
   if (moduleID.Item == SF.getParentModule()->getName() &&
       modulePath.size() == 1) {
-    if (auto importer = Context.getClangModuleLoader())
+    if (auto importer = ctx.getClangModuleLoader())
       return importer->loadModule(moduleID.Loc, modulePath);
     return nullptr;
   }
-  
-  return Context.getModule(modulePath);
+
+  return ctx.getModule(modulePath);
+}
+
+NullablePtr<ModuleDecl>
+UnboundImport::getTopLevelModule(ModuleDecl *M, SourceFile &SF) {
+  if (modulePath.size() == 1)
+    return M;
+
+  // If we imported a submodule, import the top-level module as well.
+  Identifier topLevelName = modulePath.front().Item;
+  ModuleDecl *topLevelModule = SF.getASTContext().getLoadedModule(topLevelName);
+
+  if (!topLevelModule) {
+    // Clang can sometimes import top-level modules as if they were
+    // submodules.
+    assert(!M->getFiles().empty() &&
+           isa<ClangModuleUnit>(M->getFiles().front()));
+    return M;
```

Comment:
Should this be:

```
    assert(!topLevelModule->getFiles().empty() &&
            isa<ClangModuleUnit>(topLevelModule->getFiles().front()));
     return topLevelModule;
```

or does `M == topLevelModule` here?

Reply:
Oh nvm, this is resolved by the later one.

Reply:
This function was [extracted and only lightly reformatted](https://github.com/apple/swift/blame/6616e812c6f508fc4ad4c39c7e0d977c9add289c/lib/Sema/NameBinding.cpp#L212-L230) from the previous implementation, so I'm not familiar with its full history. Looking at the blame, this particular bit was added in [this PR](https://github.com/apple/swift/pull/14673).

Basically, we used to just assert that every submodule has a top-level module. However, there's a corner case in clang where you can ask to import a submodule and get a top-level module instead. We handled that by asserting that, if the top-level module is nullptr, then the submodule is a clang module, and then returning the clang module as though it were the top-level module (which of course it actually is).